sw/device/tests/chip_power_sleep_load.c

```diff
@@ -0,0 +1,404 @@
+        dif_pwm_channel_set_enabled(&pwm, kPwmChannel[i], kDifToggleDisabled));
+    channel_config_.duty_cycle_a = kPwmDutycycle[i];
+    CHECK_DIF_OK(
+        dif_pwm_configure_channel(&pwm, kPwmChannel[i], channel_config_));
+    CHECK_DIF_OK(
+        dif_pwm_channel_set_enabled(&pwm, kPwmChannel[i], kDifToggleEnabled));
+  }
+
+  // Enable all PWM channels
+  CHECK_DIF_OK(dif_pwm_phase_cntr_set_enabled(&pwm, kDifToggleEnabled));
+
+  // PINMUX
+  for (size_t i = 0; i < PWM_PARAM_N_OUTPUTS; ++i) {
+    CHECK_DIF_OK(
+        dif_pinmux_output_select(&pinmux, kPinmuxMioOut[i], kPinmuxOutsel[i]));
+  }
+
+  LOG_INFO("PWM active");
+
+  // OTP
+
+  // Configure OTP Control to do periodic "consistency" & "integrity" checks.
+  const dif_otp_ctrl_config_t otp_ctrl_config = {
+      .check_timeout = UINT32_MAX,
+      .integrity_period_mask = 0x1,
+      .consistency_period_mask = 0x1,
+  };
+
+  CHECK_DIF_OK(dif_otp_ctrl_configure(&otp_ctrl, otp_ctrl_config));
+
+  LOG_INFO("OTP periodic checks active");
+
+  // AON Timer
+  const uint64_t kTimeTillBark = 1000;
+
+  CHECK_DIF_OK(dif_aon_timer_watchdog_stop(&aon_timer));
+  CHECK_DIF_OK(dif_aon_timer_clear_wakeup_cause(&aon_timer));
+  expected_isr_handler = wdog_irq_handler;
+  nmi_fired = false;
+  nmi_state = (dif_rv_core_ibex_nmi_state_t){0};
+  CHECK_DIF_OK(
+      dif_rv_core_ibex_enable_nmi(&rv_core_ibex, kDifRvCoreIbexNmiSourceWdog));
+  uint32_t count_cycles =
+      aon_timer_testutils_get_aon_cycles_from_us(kTimeTillBark);
+
+  if (kDeepSleep) {
+    // activate in Wakeup mode (no need for IRQ)
+    aon_timer_testutils_wakeup_config(&aon_timer, kTimeTillBark);
+  } else {
+    // activate in Watchdog mode & IRQ
+    aon_timer_testutils_watchdog_config(&aon_timer, count_cycles, UINT32_MAX,
+                                        false);
+  }
+  LOG_INFO("AON Timer active");
+
+  // Power Manager
+  dif_pwrmgr_domain_config_t pwrmgr_cfg;
+  dif_pwrmgr_request_sources_t pwrmgr_wakeups;
+
+  // Specify which request sources are enabled for wake-up
+  pwrmgr_wakeups =
+      (kDifPwrmgrWakeupRequestSourceOne | kDifPwrmgrWakeupRequestSourceTwo |
+       kDifPwrmgrWakeupRequestSourceThree | kDifPwrmgrWakeupRequestSourceFour |
+       kDifPwrmgrWakeupRequestSourceFive | kDifPwrmgrWakeupRequestSourceSix);
+
+  // Power manager configuration
+  pwrmgr_cfg = ((kCoreClkOff ? 0 : kDifPwrmgrDomainOptionCoreClockInLowPower) |
+                (kIoClkOff ? 0 : kDifPwrmgrDomainOptionIoClockInLowPower) |
+                (kUsbSlpOff ? 0 : kDifPwrmgrDomainOptionUsbClockInLowPower) |
+                (kUsbActOff ? 0 : kDifPwrmgrDomainOptionUsbClockInActivePower) |
+                (kDeepSleep ? 0 : kDifPwrmgrDomainOptionMainPowerInLowPower));
+  pwrmgr_testutils_enable_low_power(&pwrmgr, pwrmgr_wakeups, pwrmgr_cfg);
+  LOG_INFO("Power Manage configured");
+
+  CHECK_DIF_OK(dif_gpio_write(&gpio, 2, 1));
+  LOG_INFO("all HW is active");
+  test_status_set(0xff20);
+  wait_for_interrupt();
```

Comment:
Are interrupts unmasked on Ibex anywhere? It seems like this would be a never-ending sleep.

Reply:
Is there something else needed besides this? The test seems to pass on DV. 

```c
dif_rv_core_ibex_enable_nmi(&rv_core_ibex, kDifRvCoreIbexNmiSourceWdog)
```

Reply:
We can probably remove it. The test only uses it to ensure that no external interrupts are somehow triggered.

Reply:
I think WFI should not be used to wait for NMIs, at least not directly. It's impossible to atomically ascertain that the NMI has not occurred and perform WFI if it hasn't.

Instead, there would need to be a spin-wait or the use of some maskable software interrupt that could be left pending (and handled after the WFI).

Reply:
Actually, the CPU does not need an interrupt to get out of sleep mode.
It need a wake-up from Power Manager - that is what happens in this test
(and in fact, in this test, the CPU successfully wakes from both sleep and deep-sleep).

Reply:
...and, as I mentioned before, WFI is a must here. Without WFI the CPU will not go to sleep!

Reply:
Right now, the watchdog bark NMI races with WFI when we don't go into deep sleep. The handler is called before WFI, resulting in a very long sleep that makes the test time out on the FPGA.